include/streaming_ordered_output_thread_pool_executor.hpp

```diff
@@ -14,26 +14,137 @@
 #include <functional>
 #include <map>
 #include <mutex>
+#include <optional>
+#include <signal_handler.hpp>
 #include <thread>
+#include <variant>
 #include <vector>
+#if defined(__linux__)
+#include <pthread.h>
+#endif
 
-#include <signal_handler.hpp>
+struct TerminationToken {};
 
 // thread pool executing work but outputting it the order of the input
 template <typename ReturnType> class StreamingOrderedOutputThreadPoolExecutor {
 
   public:
-    StreamingOrderedOutputThreadPoolExecutor(int num_workers);
-    ~StreamingOrderedOutputThreadPoolExecutor();
+    using ReturnTypeOrTerminationToken = std::variant<ReturnType, TerminationToken>;
```

Comment:
wut. 

does termination body get some contents otherwise, questionable. 


if not think about adding: `ReturnType::Termination` or smth like this.

Reply:
I can also define it in the StreamingOrderedOutputThreadPoolExecutor<ReturnType> class.

I use it to terminate the thread that reads this output queue as an only input.

Reply:
why not take a `std::option`? 

```bash
using ReturnTypeOrTerminationToken = std::option<ReturnType>
```

Reply:
As discussed offline we will refactor the message passing system in another PR.